language-support/hs/bindings/src/DA/Ledger/Services/MeteringReportService.hs

```diff
@@ -0,0 +1,119 @@
+-- Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module DA.Ledger.Services.MeteringReportService (
+    getMeteringReport, MeteringReport(..), iso8601ToTimestamp
+    ) where
+
+
+import Data.Aeson ( KeyValue((.=)), ToJSON(..), object)
+import DA.Ledger.Convert
+import DA.Ledger.GrpcWrapUtils
+import DA.Ledger.LedgerService
+import DA.Ledger.Types
+import qualified Data.Text.Lazy as TL
+import Network.GRPC.HighLevel.Generated
+import qualified Com.Daml.Ledger.Api.V1.Admin.MeteringReportService as LL
+import Data.Maybe (fromMaybe, maybeToList)
+import qualified Data.Time.Clock as Clock
+import qualified Data.Time.Clock.System as System
+import qualified Data.Time.Format.ISO8601 as ISO8601
+import GHC.Int (Int64)
+import GHC.Word (Word32)
+
+data MeteredApplication = MeteredApplication {
+  application :: ApplicationId
+, events :: Int64
+} deriving (Show)
+
+instance ToJSON MeteredApplication where
+  toJSON (MeteredApplication application events) =
+    object
+    [   "application" .= (unApplicationId application)
+    ,   "events" .= events
+    ]
+
+data MeteringReport = MeteringReport {
+  participant :: ParticipantId
+, from  :: Maybe Timestamp
+, toRequested :: Maybe Timestamp
+, toActual :: Timestamp
+, applications :: [MeteredApplication]
+} deriving (Show)
+
+instance ToJSON MeteringReport where
+  toJSON (MeteringReport participant from toRequested toActual applications) =
+    object (
+    [   "participant" .= (unParticipantId participant)
+    ,   "to-actual" .= (timestampToIso8601 toActual)
+    ,   "applications" .= applications
+    ]
+    ++ maybeToList (fmap (("from" .=) . timestampToIso8601) from)
+    ++ maybeToList (fmap (("to-requested" .=) . timestampToIso8601) toRequested)
+    )
+
+timestampToSystemTime :: Timestamp -> System.SystemTime
+timestampToSystemTime ts = st
+  where
+    s = (fromIntegral (seconds ts) :: Int64)
+    n = (fromIntegral (nanos ts) :: Word32)
+    st = System.MkSystemTime s n
+
+systemTimeToTimestamp :: System.SystemTime -> Timestamp
+systemTimeToTimestamp st = ts
+  where
+    s = (fromIntegral (System.systemSeconds st) :: Int)
+    n = (fromIntegral (System.systemNanoseconds st) :: Int)
+    ts = Timestamp s n
+
+timestampToIso8601 :: Timestamp -> String
```

Comment:
I don’t think we should be converting between utc and system time. ledgers run in utc, let users specify times in utc as well.

Reply:
Users do provide a UTC time in ISO 8601 format.  The SystemTime is an implementation detail as it appeared to be the most convenient way to get between `Timestamp` and `UTCTime`.

Reply:
sorry I confused SystemTime with LocalTime, all good :+1: